proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java

```diff
@@ -518,15 +521,14 @@ public void handle(final EncryptionResponse encryptResponse) throws Exception
 
         String encName = URLEncoder.encode( InitialHandler.this.getName(), "UTF-8" );
 
-        MessageDigest sha = MessageDigest.getInstance( "SHA-1" );
-        for ( byte[] bit : new byte[][]
-        {
-            request.getServerId().getBytes( "ISO_8859_1" ), sharedKey.getEncoded(), EncryptionUtil.keys.getPublic().getEncoded()
-        } )
-        {
-            sha.update( bit );
-        }
-        String encodedHash = URLEncoder.encode( new BigInteger( sha.digest() ).toString( 16 ), "UTF-8" );
+        MessageDigest sha = MessageDigest.getInstance("SHA-1");
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+        buffer.put(request.getServerId().getBytes(StandardCharsets.ISO_8859_1));
+        buffer.put(sharedKey.getEncoded());
+        buffer.put(EncryptionUtil.keys.getPublic().getEncoded());
+        sha.update(buffer.array());
```

Comment:
The array method returns the underlying byte[] of the buffer, its length is the buffer's capacity, so too long, need to use sth like `update(buffer.array(), 0, buffer.position())`
or maybe don't use the fancy buffer class, and allocate byte[] array with correct length and manually call System.arraycopy 3 times.